### PR TITLE
Reject connection attempts while closing

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today if there is a constant stream of connection attempts then it's possible for the `ClusterConnectionManager` to wait forever in `close()` for `connectingRefCounter` to be fully released. With this commit we reject connection attempts while closing, avoiding this starvation situation.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Spotted this while looking at a test failure which apparently failed due to getting stuck exactly here: https://gradle-enterprise.elastic.co/s/rdlucvijy2dby. I'm not sure this is actually the reason for this failure, it is kinda implausible that we open connections at a high enough rate to cause this starvation, but still I don't see any other obvious reasons and this is worth fixing.

**Review by pxsalehi — APPROVED** [review]
LGTM. But a small nit: the test setup seems not very straight forward. Individual pieces/blocks of it could benefit from some extra comments.

**Review by henningandersen — APPROVED** [review]
LGTM.

File: server/src/test/java/org/elasticsearch/transport/ClusterConnectionManagerTests.java [hunk 1]
```diff
@@ -346,6 +350,107 @@ public void testConcurrentConnects() throws Exception {
...
+                final var listener = new ActionListener<Releasable>() {
+                    @Override
+                    public void onResponse(Releasable releasable) {
+                        releasable.close();
+                        threadPool.generic().execute(ConnectionLoop.this);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        assertThat(e, instanceOf(ConnectTransportException.class));
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/transport/ClusterConnectionManagerTests.java:409]: :+1: yes good catch, see 92bb6e57e389475ceacf5e202d417a26a9160213

**Comment by DaveCTurner** [comment]
> LGTM. But a small nit: the test setup seems not very straight forward. Individual pieces/blocks of it could benefit from some extra comments.

Fair comment :) I added some more detail in 2c3b8c600d79e1d8fdfce1344228aad601c0e940, hope that helps.

**Comment by DaveCTurner** [comment]
On reflection I think this might fix the test failure mentioned above after all. We don't have to constantly open connections until the suite times out, we only have to do it until we terminate the threadpool, because we're using `GENERIC` threads here which just silently drop work on the floor at shutdown and therefore could well leak a ref.

**Pull request merged**